primary/src/grpc_server/proposer.rs

```diff
@@ -0,0 +1,61 @@
+use crate::grpc_server::public_key_mapper::PublicKeyMapper;
+use consensus::dag::Dag;
+use crypto::traits::VerifyingKey;
+use std::sync::Arc;
+use tonic::{Request, Response, Status};
+use types::{Proposer, RoundsRequest, RoundsResponse};
+
+pub struct NarwhalProposer<PublicKey: VerifyingKey, KeyMapper: PublicKeyMapper<PublicKey>> {
+    dag: Option<Arc<Dag<PublicKey>>>,
+
+    public_key_mapper: KeyMapper,
+}
+
+impl<PublicKey: VerifyingKey, KeyMapper: PublicKeyMapper<PublicKey>>
+    NarwhalProposer<PublicKey, KeyMapper>
+{
+    pub fn new(dag: Option<Arc<Dag<PublicKey>>>, public_key_mapper: KeyMapper) -> Self {
+        Self {
+            dag,
+            public_key_mapper,
+        }
+    }
+}
+
+#[tonic::async_trait]
+impl<PublicKey: VerifyingKey, KeyMapper: PublicKeyMapper<PublicKey>> Proposer
+    for NarwhalProposer<PublicKey, KeyMapper>
+{
+    /// Retrieves the min & max rounds that contain collections available for
+    /// block proposal for the dictated validator.
+    /// by the provided public key.
+    async fn rounds(
+        &self,
+        request: Request<RoundsRequest>,
+    ) -> Result<Response<RoundsResponse>, Status> {
+        // convert key
+        let key = self
+            .public_key_mapper
+            .map(
+                request
+                    .into_inner()
+                    .public_key
+                    .ok_or_else(|| Status::invalid_argument("Not public key provided"))?,
+            )
+            .map_err(|_| Status::invalid_argument("Couldn't parse provided key"))?;
```

Comment:
I am having this practical problem of converting a `PublicKeyProto` to an `Ed25519PublicKey`. Since the `NarwhalProposer` accepts the public key as a generic type argument (`<PublicKey: VerifyingKey>`) , is not possible to just parse at this point the key directly like:
```
let key = Ed25519PublicKey::from_bytes(public_key_proto.bytes.as_ref())
```

when I try to use the key on the dag (e.x `dag.rounds(key)`) it complains - correctly - because the `dag` is accepting the generic type `PublicKey` as type. 

Given the above the only solution I could come up with was introducing a generic mapper trait `PublicKeyMapper` and inject all the way up from the node initialisation the concrete implementation of it, the `Ed25519PublicKeyMapper`. That being said, I don't find it very pretty but I don't have another idea of how this could be achieved alternatively. In the same time I have the feeling that I am missing something obvious - thoughts @huitseeker @arun-koshy ?

Reply:
1. Did you notice https://github.com/MystenLabs/narwhal/pull/244/files#diff-8efd1201ab5d436d6cf9a7dbe1a579923c267371eada8742705244a670317400R38-R53 ?
2. This will disappear when I tackle https://github.com/MystenLabs/narwhal/issues/246

Reply:
1. Yes , but it's basically the same thing - more specifically the `impl TryFrom<&PublicKeyProto> for Ed25519PublicKey` will try to convert the input to a `Ed25519PublicKey`, when the `dag` on the `NarwhalProposer` is passed with the generic type `PublicKey`. So same result.
2. Ah yeah, that would definitely help on this!